### ISSUE TITLE
mBERT Inference Speed for Language Modelling

### ISSUE THREAD

**Issue by giovgiac** [aws/aws-neuron-sdk#397]
Hello!

First off, thank you for the great library and cost-effective instance type made available through AWS. We're quite looking forward to using it in some of our services at the company I work at.

Currently, I'm attempting to optimize [BertForMaskedLM](https://huggingface.co/docs/transformers/model_doc/bert#transformers.BertForMaskedLM) (from Transformers). I managed to get it compiled properly and it even executes correctly, however it ends up being many times slower than just executing it with Transformers on the inf1.2xlarge instance. Also, when I look at RAM consumption, it seems to consume a pretty hefty amount, definitely more than when executing it normally. So I was wondering if there's anything I'm doing wrong or is it some problem that Neuron has with optimizing the Masked LM Head?

This is the code I'm currently using to compile the model (copied from my Jupyter Notebook):

```python
import time
import torch
import torch_neuron

from transformers import BertForMaskedLM, BertTokenizer


class BertForNeuron(torch.nn.Module):
    def __init__(self, model):
        super().__init__()
        self.model = model
        
    def forward(self, input_ids, attention_mask, token_type_ids, labels):
        loss, logits = self.model(
            input_ids,
            attention_mask=attention_mask,
            token_type_ids=token_type_ids,
            labels=labels,
            return_dict=False
        )
        
        return loss, logits


model = BertForMaskedLM.from_pretrained("bert-base-multilingual-cased", return_dict=False)
neuron_model = BertForNeuron(model)
tokenizer = BertTokenizer.from_pretrained("bert-base-multilingual-cased")

batch_size = 1
max_length = 256

masked = "Hello, my friend! How [MASK] you doing today?"
sentence = "Hello, my friend! How are you doing today?"

inputs = tokenizer.batch_encode_plus(batch_size * [masked], max_length=max_length, padding="max_length", truncation=True, return_tensors="pt")
targets = tokenizer.batch_encode_plus(batch_size * [sentence], max_length=max_length, padding="max_length", truncation=True, return_tensors="pt")

example_inputs = inputs["input_ids"], inputs["attention_mask"], inputs["token_type_ids"], targets["input_ids"]

# Warm-up
neuron_model(*example_inputs)

stime = time.time()
org_outputs = neuron_model(*example_inputs)
print(time.time() - stime)

optimized = torch_neuron.trace(neuron_model, example_inputs)

# Warm-up
optimized(*example_inputs)

stime = time.time()
outputs = optimized(*example_inputs)
print(time.time() - stime)

optimized.save("neuron-mbert.pt")
```

**Comment by jeffhataws** [comment]
Thank you @giovgiac for filing this issue. We will take a look and get back to you.

**Comment by micwade-aws** [comment]
@giovgiac - We are able to recreate the issue and it's due to the large embedding table, which neuron-cc will map to the CPU.  What we're still working on is a profile to understand why, despite this, we're not performing better than CPU-only execution.

**Comment by hannanjgaws** [comment]
HI @giovgiac,

Manually placing the `aten::embedding` operator on Neuron and using a wrapper that runs `CrossEntropyLoss` and the output format manipulation code on CPU should improve performance of your model.

The following script demonstrates how to write two new wrappers, `mBERTNeuron` and `mBERTWrapper`, to improve the performance of your model:
- `BertForNeuron` extracts the compute heavy part of your model and compiles it for optimized inference on Neuron
- `mBERTWrapper` takes in the compiled `BertForNeuron` model and runs the `CrossEntropyLoss` and output format manipulation code on CPU. The `mBERTWrapper` model is the “full” end-to-end model that you will want to use for inference

You can use the example `trace` function to compile your model using the wrappers with different batch sizes. Please make sure to run the `trace` compilation function on a large compute instance to avoid OOM errors during compilation.

Additionally, the `infer` function compares the performance of the new `mBERTWrapper` to the performance the old `BertForNeuron` CPU model. Please note that you can use any inf1 instance size (`inf1.xlarge`, `inf1.2xlarge`, `inf6.xlarge`, etc.) for inference testing and deployment.

Please let us know if these scripts resolve the performance issues you were observing on inf1.

[mbert_neuron.txt](https://github.com/aws/aws-neuron-sdk/files/8755391/mbert_neuron.txt)

**Comment by giovgiac** [comment]
Hello @hannanjgaws!

Thank you very much for looking into the issue. I'll test these scripts and get back to you 😃 !

**Comment by giovgiac** [comment]
@hannanjgaws Just confimed, it indeed runs faster than CPU-only! Thank you very much for your help 🙇🏻‍♂️
